typescript/sdk/src/token/CosmosNativeWarpModule.ts

```diff
@@ -0,0 +1,469 @@
+
+import { CosmosNativeWarpRouteReader } from './CosmosNativeWarpRouteReader.js';
+import { TokenType } from './config.js';
+import { HypTokenRouterConfig, HypTokenRouterConfigSchema } from './types.js';
+
+type WarpRouteAddresses = {
+  deployedTokenRoute: Address;
+};
+
+export class CosmosNativeWarpModule extends HyperlaneModule<
+  ProtocolType.CosmosNative,
+  HypTokenRouterConfig,
+  WarpRouteAddresses
+> {
+  protected logger = rootLogger.child({
+    module: 'CosmosNativeWarpModule',
+  });
+  reader: CosmosNativeWarpRouteReader;
+  public readonly chainName: ChainName;
+  public readonly chainId: string;
+  public readonly domainId: Domain;
+
+  constructor(
+    protected readonly metadataManager: ChainMetadataManager,
+    args: HyperlaneModuleParams<HypTokenRouterConfig, WarpRouteAddresses>,
+    protected readonly signer: SigningHyperlaneModuleClient,
+  ) {
+    super(args);
+    this.reader = new CosmosNativeWarpRouteReader(
+      metadataManager,
+      args.chain,
+      signer,
+    );
+    this.chainName = this.metadataManager.getChainName(args.chain);
+    this.chainId = metadataManager.getChainId(args.chain).toString();
+    this.domainId = metadataManager.getDomainId(args.chain);
+  }
+
+  /**
+   * Retrieves the token router configuration for the specified address.
+   *
+   * @param address - The address to derive the token router configuration from.
+   * @returns A promise that resolves to the token router configuration.
+   */
+  async read(): Promise<HypTokenRouterConfig> {
+    return this.reader.deriveWarpRouteConfig(
+      this.args.addresses.deployedTokenRoute,
+    );
+  }
+
+  /**
+   * Updates the Warp Route contract with the provided configuration.
+   *
+   * @param expectedConfig - The configuration for the token router to be updated.
+   * @returns An array of Cosmos transactions that were executed to update the contract, or an error if the update failed.
+   */
+  async update(
+    expectedConfig: HypTokenRouterConfig,
+  ): Promise<AnnotatedCosmJsNativeTransaction[]> {
+    HypTokenRouterConfigSchema.parse(expectedConfig);
+    const actualConfig = await this.read();
+
+    const transactions = [];
+
+    /**
+     * @remark
+     * The order of operations matter
+     * 1. createOwnershipUpdateTxs() must always be LAST because no updates possible after ownership transferred
+     * 2. createRemoteRoutersUpdateTxs() must always be BEFORE createSetDestinationGasUpdateTxs() because gas enumeration depends on domains
+     */
+    transactions.push(
+      ...(await this.createIsmUpdateTxs(actualConfig, expectedConfig)),
+      ...this.createEnrollRemoteRoutersUpdateTxs(actualConfig, expectedConfig),
+      ...this.createUnenrollRemoteRoutersUpdateTxs(
+        actualConfig,
+        expectedConfig,
+      ),
+      ...(await this.createSetDestinationGasUpdateTxs(
+        actualConfig,
+        expectedConfig,
+      )),
+      ...this.createOwnershipUpdateTxs(actualConfig, expectedConfig),
```

Comment:
No hook update?

Reply:
currently there are no hooks in tokens yet on cosmos native, this could change with future releases but therefore for now no hooks updates are needed